File: Python Tasks/Algorithms.py
import random


# Searching Algorithms:

from random import randint
from turtle import pos, position


def linearSearch(list, target):
    # Not using For / For Each loop:
    index = 0
    while index < len(list):
        if (list[index] == target): return index
        index += 1
    return -1

def binarySearch(list, startIndex, endIndex, target):
    # startIndex and endIndex are redundant for the first call
    midpoint = (startIndex + endIndex) // 2

    if (list[midpoint] > target):
        return binarySearch(list, startIndex, midpoint - 1, target)
    elif (list[midpoint] < target):
        return binarySearch(list, midpoint + 1, endIndex, target)    
    else:
        return midpoint

# Sorting Algorithms:

def bubbleSort(list):
    done = False
    while not done:
        done = True
        index = 0
        while (index < len(list) - 1):
            
            firstItem = list[index]
            secondItem = list[index + 1]

            if firstItem > secondItem:
                # Swapping items
                list[index] = secondItem
                list[index + 1] = firstItem
                done = False

                print(list)
            index += 1 
      
    return list   

def insertionSort(list):
    index = 1

    while (index < len(list)):
        currentValue = list[index]
        position = index
        while (position > 0 and currentValue < list[position - 1]):
            # Shifting all values
            list[position] = list[position - 1]
            position -= 1

        # Inserting
        list[position] = currentValue
        index += 1      

    return list            


def mergeSort(list):
    pass


def quickSort():
    pass


# Generates a random list of integers of the specified size:
def randomList(length, ordered):
    list = []
    for index in range(length):
        list.append(random.randint(0, 100))
    
    if (ordered): return insertionSort(list)
    return list

File: Algorithms/Algorithms.java

class Algorithms{

    public static void main(String[] args) {
        // The main method is called when the program is run:
        Algorithms algorithms = new Algorithms();

        int[] values = new int[]{1, 4, 3, 5, 6, 3, 4, 6, 4};
        values = algorithms.bubbleSort(values);

        for (int value: values) {
            System.out.println(value);
        }
    }

    public int linearSearch(String[] values, String target){
        int index = 0;

        while (index < values.length) {
            if (values[index].equals(target)){
                return index;
            }
            index++;
        }
        return -1;
}

    public int recursiveBinarySearch(int[] values, int target, int start, int end){
        int midpoint = (start + end) / 2;

        if (start > end){return -1;}
        else if (target < values[midpoint]){return recursiveBinarySearch(values, target, start, midpoint-1);}
        else if (values[midpoint] < target){return recursiveBinarySearch(values, target, midpoint+1, end);}
        
        return midpoint;
    }

    public int iterativeBinarySearch(int[] values, int target){
        int midpoint;
        int start = 0;
        int end = values.length - 1;

        while (start <= end){
            midpoint = (start + end) / 2;
            
            if (target < values[midpoint]){end = midpoint - 1;}
            else if (values[midpoint] < target){start = midpoint + 1;}
            else {return midpoint;}
        }
        return -1;
    }

    public int[] bubbleSort(int[] values){
        int temp;
        boolean sorted = false;

        while (!sorted) {
            sorted = true;

            for (int index=0; index<values.length-1; index++){
                if (values[index] > values[index+1]){
                    sorted = false;

                    temp = values[index];
                    values[index] = values[index+1];
                    values[index+1] = temp;
                }
            }
        }
        return values;
    }

    public int[] insertionSort(int[] values){
        int index = 1;
        int position;
        int currentValue;

        while (index < values.length){
            currentValue = values[index];
            position = index;

            while (position > 0 && currentValue < values[position - 1]){
                values[position] = values[position - 1];
                position --;
            }
            values[position] = currentValue;
            index ++;
        }

        return values;
    }

    }
